File: python/x402/mechanisms/evm/batch_settlement/server/settle.py
"""Server-side settle lifecycle hooks for the batch-settlement EVM scheme."""

from __future__ import annotations

import re
import time
from typing import TYPE_CHECKING, Any

from ..authorizer_signer import sign_claim_batch, sign_refund
from ..errors import (
    ERR_CHANNEL_BUSY,
    ERR_CHARGE_EXCEEDS_SIGNED_CUMULATIVE,
    ERR_CUMULATIVE_AMOUNT_MISMATCH,
    ERR_INVALID_VOUCHER_SIGNATURE,
    ERR_MISSING_CHANNEL,
    ERR_REFUND_AMOUNT_INVALID,
    ERR_REFUND_NO_BALANCE,
)
from ..types import (
    VoucherClaim,
    is_deposit_payload,
    is_refund_payload,
    is_voucher_payload,
)
from ..utils import compute_channel_id
from .storage import Channel
from .utils import (
    parse_refund_settlement_snapshot,
    read_channel_state_extra,
    read_extra_number,
    read_extra_string,
)

if TYPE_CHECKING:
    from .....schemas.hooks import (
        SettleContext,
        SettleFailureContext,
        SettleResultContext,
    )
    from .scheme import BatchSettlementEvmScheme


_UINT_RE = re.compile(r"^\d+$")


def _now_ms() -> int:
    return int(time.time() * 1000)


def _channel_state_extra(
    channel: Channel, charged_cumulative_amount: str | None = None
) -> dict[str, Any]:
    out: dict[str, Any] = {
        "channelId": channel.channel_id,
        "balance": channel.balance,
        "totalClaimed": channel.total_claimed,
        "withdrawRequestedAt": channel.withdraw_requested_at,
        "refundNonce": str(channel.refund_nonce),
    }
    if charged_cumulative_amount is not None:
        out["chargedCumulativeAmount"] = charged_cumulative_amount
    return out


def handle_before_settle(scheme: BatchSettlementEvmScheme, ctx: SettleContext):
    """For voucher payloads: short-circuit settle by incrementing local charged total.

    Cooperative refund and deposit payloads fall through to the facilitator.
    """
    from .....schemas import SettleResponse
    from .....schemas.hooks import AbortResult, SkipSettleResult

    payment_payload = ctx.payment_payload
    requirements = ctx.requirements
    raw = payment_payload.payload

    if not is_voucher_payload(raw):
        return None

    voucher = raw["voucher"]
    channel_id = voucher["channelId"]
    pending_id = None
    request_context = scheme.read_request_context(payment_payload)
    if request_context is not None:
        pending_id = request_context.pending_id

    increment = int(requirements.amount)
    signed_cap = int(voucher["maxClaimableAmount"])
    outcome: dict[str, Any] = {}

    def update(current: Channel | None) -> Channel | None:
        if current is None:
            outcome["status"] = "missing"
            return current
        if not pending_id or (
            current.pending_request is None or current.pending_request.pending_id != pending_id
        ):
            outcome["status"] = "pending_mismatch"
            return current

        new_charged = int(current.charged_cumulative_amount) + increment
        if new_charged > signed_cap:
            outcome["status"] = "cap_exceeded"
            outcome["charged"] = str(new_charged)
            next_ch = current.copy()
            next_ch.pending_request = None
            return next_ch

        next_ch = current.copy()
        next_ch.charged_cumulative_amount = str(new_charged)
        next_ch.signed_max_claimable = str(voucher["maxClaimableAmount"])
        next_ch.signature = voucher.get("signature", current.signature)
        next_ch.last_request_timestamp = _now_ms()
        next_ch.pending_request = None
        outcome["status"] = "committed"
        outcome["previous"] = current
        outcome["current"] = next_ch
        return next_ch

    update_result = scheme.get_storage().update_channel(channel_id, update)

    status = outcome.get("status")
    if status == "missing":
        scheme.take_request_context(payment_payload)
        return AbortResult(reason=ERR_MISSING_CHANNEL)

    if status == "cap_exceeded":
        scheme.take_request_context(payment_payload)
        return AbortResult(reason=ERR_CHARGE_EXCEEDS_SIGNED_CUMULATIVE)

    if update_result.status != "updated" or status != "committed":
        scheme.take_request_context(payment_payload)
        return AbortResult(reason=ERR_CHANNEL_BUSY)

    scheme.take_request_context(payment_payload)

    previous: Channel = outcome["previous"]
    current_ch: Channel = outcome["current"]
    skip_extra: dict[str, Any] = {
        "channelState": _channel_state_extra(previous, current_ch.charged_cumulative_amount),
        "chargedAmount": requirements.amount,
    }
    response = SettleResponse(
        success=True,
        payer=previous.channel_config.payer.lower(),
        transaction="",
        network=requirements.network,
        amount="",
        extra=skip_extra,
    )
    return SkipSettleResult(result=response)


def handle_enrich_settlement_payload(
    scheme: BatchSettlementEvmScheme, ctx: SettleContext
) -> dict[str, Any] | None:
    """For cooperative refund payloads: attach refund + claim batch + authorizer sigs."""
    payment_payload = ctx.payment_payload
    requirements = ctx.requirements
    raw = payment_payload.payload
    if not is_refund_payload(raw):
        return None

    network = str(requirements.network)
    from ..types import ChannelConfig

    channel_config = ChannelConfig.from_dict(raw["channelConfig"])
    channel_id = compute_channel_id(channel_config, network)
    if raw["voucher"]["channelId"] != channel_id:
        raise ValueError("refund channelId does not match channelConfig")

    channel = scheme.get_storage().get(channel_id)
    if channel is None:
        raise ValueError(ERR_MISSING_CHANNEL)

    pending_id = None
    request_context = scheme.read_request_context(payment_payload)
    if request_context is not None:
        pending_id = request_context.pending_id
    if not pending_id or (
        channel.pending_request is None or channel.pending_request.pending_id != pending_id
    ):
        raise ValueError(ERR_CHANNEL_BUSY)

    if int(raw["voucher"]["maxClaimableAmount"]) != int(channel.charged_cumulative_amount):
        raise ValueError(ERR_CUMULATIVE_AMOUNT_MISMATCH)
    if raw["voucher"].get("signature") != channel.signature:
        raise ValueError(ERR_INVALID_VOUCHER_SIGNATURE)

    claim_entry = VoucherClaim(
        channel=channel_config,
        max_claimable_amount=str(raw["voucher"]["maxClaimableAmount"]),
        signature=raw["voucher"]["signature"],
        total_claimed=channel.charged_cumulative_amount,
    )

    remainder = int(channel.balance) - int(channel.charged_cumulative_amount)
    if remainder <= 0:
        raise ValueError(ERR_REFUND_NO_BALANCE)

    refund_amount_int = remainder
    raw_amount = raw.get("amount")
    if raw_amount is not None:
        if not _UINT_RE.match(str(raw_amount)):
            raise ValueError(ERR_REFUND_AMOUNT_INVALID)
        requested = int(raw_amount)
        if requested <= 0:
            raise ValueError(ERR_REFUND_AMOUNT_INVALID)
        if requested > remainder:
            raise ValueError(ERR_REFUND_NO_BALANCE)
        refund_amount_int = requested

    refund_amount = str(refund_amount_int)
    nonce = str(channel.refund_nonce or 0)

    signer = scheme.get_receiver_authorizer_signer()
    refund_authorizer_signature = (
        sign_refund(signer, channel_id, refund_amount, nonce, network)
        if signer is not None
        else None
    )
    claim_authorizer_signature = (
        sign_claim_batch(signer, [claim_entry], network) if signer is not None else None
    )

    scheme.remember_channel_snapshot(payment_payload, channel)

    out: dict[str, Any] = {
        "refundNonce": nonce,
        "claims": [claim_entry.to_dict()],
    }
    if raw_amount is None:
        out["amount"] = refund_amount
    if refund_authorizer_signature is not None:
        out["refundAuthorizerSignature"] = refund_authorizer_signature
    if claim_authorizer_signature is not None:
        out["claimAuthorizerSignature"] = claim_authorizer_signature
    return out


def handle_after_settle(scheme: BatchSettlementEvmScheme, ctx: SettleResultContext) -> None:
    """Update channel state after facilitator settle for refund / deposit payloads."""
    from ..types import ChannelConfig

    payment_payload = ctx.payment_payload
    requirements = ctx.requirements
    result = ctx.result
    if not result.success:
        return

    raw = payment_payload.payload
    storage = scheme.get_storage()

    if is_refund_payload(raw):
        channel_config = ChannelConfig.from_dict(raw["channelConfig"])
        channel_id = compute_channel_id(channel_config, str(requirements.network))
        request_context = scheme.read_request_context(payment_payload)
        pending_id = request_context.pending_id if request_context else None
        now = _now_ms()
        snapshot = parse_refund_settlement_snapshot(result.extra)

        def update_refund(current: Channel | None) -> Channel | None:
            if current is None:
                return current
            if not pending_id or (
                current.pending_request is None or current.pending_request.pending_id != pending_id
            ):
                return current
            if int(snapshot.balance) <= int(current.charged_cumulative_amount):
                # Channel fully drained — delete the record.
                return None
            next_ch = current.copy()
            next_ch.balance = snapshot.balance
            next_ch.total_claimed = snapshot.total_claimed
            next_ch.withdraw_requested_at = snapshot.withdraw_requested_at
            next_ch.refund_nonce = snapshot.refund_nonce
            next_ch.onchain_synced_at = now
            next_ch.last_request_timestamp = now
            next_ch.pending_request = None
            return next_ch

        update_result = storage.update_channel(channel_id, update_refund)
        if update_result.status == "unchanged":
            raise ValueError(ERR_CHANNEL_BUSY)
        return

    if is_voucher_payload(raw):
        return

    if is_deposit_payload(raw):
        voucher = raw["voucher"]
        channel_id = voucher["channelId"]
        channel_config = ChannelConfig.from_dict(raw["channelConfig"])
        request_context = scheme.read_request_context(payment_payload)
        pending_id = request_context.pending_id if request_context else None
        extra = result.extra or {}
        channel_state = read_channel_state_extra(extra)
        signed_max_claimable = str(voucher["maxClaimableAmount"])
        signature = voucher.get("signature", "0x")
        now = _now_ms()

        def update_deposit(current: Channel | None) -> Channel | None:
            if current is None:
                return current
            if not pending_id or (
                current.pending_request is None or current.pending_request.pending_id != pending_id
            ):
                return current
            charged_actual = str(int(current.charged_cumulative_amount) + int(requirements.amount))
            next_ch = current.copy()
            next_ch.channel_config = channel_config
            next_ch.charged_cumulative_amount = charged_actual
            next_ch.signed_max_claimable = signed_max_claimable
            next_ch.signature = signature
            next_ch.balance = read_extra_string(channel_state, "balance", current.balance)
            next_ch.total_claimed = read_extra_string(
                channel_state, "totalClaimed", current.total_claimed
            )
            next_ch.withdraw_requested_at = read_extra_number(
                channel_state, "withdrawRequestedAt", current.withdraw_requested_at
            )
            next_ch.refund_nonce = read_extra_number(
                channel_state, "refundNonce", current.refund_nonce
            )
            next_ch.onchain_synced_at = now
            next_ch.last_request_timestamp = now
            next_ch.pending_request = None
            return next_ch

        update_result = storage.update_channel(channel_id, update_deposit)
        if update_result.status == "updated" and update_result.channel is not None:
            scheme.remember_channel_snapshot(payment_payload, update_result.channel)
            return
        scheme.take_request_context(payment_payload)
        raise ValueError(ERR_CHANNEL_BUSY)


def handle_settle_failure(scheme: BatchSettlementEvmScheme, ctx: SettleFailureContext) -> None:
    """Clear this request's reservation after settlement throws."""
    scheme.clear_pending_request(ctx.payment_payload)


def handle_enrich_settlement_response(
    scheme: BatchSettlementEvmScheme, ctx: SettleResultContext
) -> dict[str, Any] | None:
    """Supply server-owned settlement response fields from the channel snapshot."""
    raw = ctx.payment_payload.payload
    if is_voucher_payload(raw):
        return None

    channel = scheme.take_channel_snapshot(ctx.payment_payload)
    if channel is None:
        return None

    if is_refund_payload(raw):
        return {
            "channelState": {
                "chargedCumulativeAmount": channel.charged_cumulative_amount,
            }
        }
    if is_deposit_payload(raw):
        return {
            "channelState": {
                "chargedCumulativeAmount": channel.charged_cumulative_amount,
            },
            "chargedAmount": ctx.requirements.amount,
        }
    return {
        "channelState": {
            "chargedCumulativeAmount": channel.charged_cumulative_amount,
        }
    }


__all__ = [
    "handle_before_settle",
    "handle_enrich_settlement_payload",
    "handle_after_settle",
    "handle_settle_failure",
    "handle_enrich_settlement_response",
]

File: typescript/packages/mechanisms/evm/src/batch-settlement/server/settle.ts
import type { SettleResponse } from "@x402/core/types";
import type { SettleContext, SettleFailureContext, SettleResultContext } from "@x402/core/server";
import { signClaimBatch, signRefund } from "../authorizerSigner";
import {
  isBatchSettlementDepositPayload,
  isBatchSettlementRefundPayload,
  isBatchSettlementVoucherPayload,
} from "../types";
import type { BatchSettlementPaymentResponseExtra, BatchSettlementVoucherClaim } from "../types";
import { computeChannelId } from "../utils";
import * as Errors from "../errors";
import type { BatchSettlementEvmScheme } from "./scheme";
import type { Channel } from "./storage";
import { rethrowLockImplementationError } from "./storage";
import {
  parseRefundSettlementSnapshot,
  readChannelStateExtra,
  readExtraNumber,
  readExtraString,
} from "./utils";

type AdmissionHold = "self" | "other" | "none";

/**
 * Inspects the admission lock for this request.
 *
 * Used for deposit/refund holder checks and snapshot recovery. This request
 * proceeds when it holds the lock or no lock is present (lost/expired).
 * Lock-store I/O failures are optimistic (`none`). Implementation/parse
 * errors fail closed and propagate to the caller.
 *
 * @param scheme - Owning scheme for lock-store access.
 * @param channelId - Channel to inspect.
 * @param pendingId - This request's lock owner, if any.
 * @returns `self` when `pendingId` holds; `other` when a different holder is live; `none` otherwise.
 */
async function inspectAdmission(
  scheme: BatchSettlementEvmScheme,
  channelId: string,
  pendingId: string | undefined,
): Promise<AdmissionHold> {
  try {
    const locks = scheme.getLockStorage();
    if (pendingId && (await locks.isHeld(channelId, pendingId))) {
      return "self";
    }
    return (await locks.isHeld(channelId)) ? "other" : "none";
  } catch (err) {
    rethrowLockImplementationError(err);
    return "none";
  }
}

/**
 * Converts stored channel state into the public response snapshot shape.
 *
 * @param channel - Stored channel state.
 * @param chargedCumulativeAmount - Optional current charged cumulative amount.
 * @returns Response-ready channel snapshot.
 */
function channelStateExtra(
  channel: Pick<
    Channel,
    "channelId" | "balance" | "totalClaimed" | "withdrawRequestedAt" | "refundNonce"
  >,
  chargedCumulativeAmount?: string,
): NonNullable<BatchSettlementPaymentResponseExtra["channelState"]> {
  return {
    channelId: channel.channelId as `0x${string}`,
    balance: channel.balance,
    totalClaimed: channel.totalClaimed,
    withdrawRequestedAt: channel.withdrawRequestedAt,
    refundNonce: String(channel.refundNonce),
    ...(chargedCumulativeAmount !== undefined ? { chargedCumulativeAmount } : {}),
  };
}

/**
 * Lifecycle hook: runs before the facilitator settles a payment.
 *
 * Voucher payloads increment `chargedCumulativeAmount` locally and return `skip` so
 * the middleware responds without an onchain settle. Refund and deposit payloads
 * fall through to facilitator settlement; their durable rows update in `afterSettle`.
 *
 * @param scheme - Owning `BatchSettlementEvmScheme` instance for storage access.
 * @param ctx - Settle lifecycle context (payload and requirements).
 * @returns Nothing to proceed; `abort` to fail; `skip` with a result to short-circuit settlement.
 */
export async function handleBeforeSettle(
  scheme: BatchSettlementEvmScheme,
  ctx: SettleContext,
): Promise<
  void | { abort: true; reason: string; message?: string } | { skip: true; result: SettleResponse }
> {
  const { paymentPayload, requirements } = ctx;

  const raw = paymentPayload.payload;
  const storage = scheme.getStorage();

  if (!isBatchSettlementVoucherPayload(raw)) {
    return;
  }

  const { voucher } = raw;
  const channelId = voucher.channelId;
  const requestContext = scheme.readRequestContext(paymentPayload);
  const snapshot = requestContext?.channelSnapshot;
  const localVerify = requestContext?.localVerify === true;
  const now = Date.now();

  const increment = BigInt(requirements.amount);
  const signedCap = BigInt(voucher.maxClaimableAmount);
  let outcome:
    | { status: "missing" }
    | { status: "cap_exceeded"; charged: string }
    | { status: "committed"; previous: Channel; current: Channel }
    | undefined;

  const charge = (current: Channel | undefined, base: Channel | undefined): Channel | undefined => {
    if (!base) {
      outcome = { status: "missing" };
      return current;
    }

    const newCharged = BigInt(base.chargedCumulativeAmount) + increment;
    if (newCharged > signedCap) {
      outcome = { status: "cap_exceeded", charged: newCharged.toString() };
      return current;
    }

    const updatedChannel: Channel = {
      ...base,
      ...(localVerify || !snapshot
        ? {}
        : {
            balance: snapshot.balance,
            totalClaimed: snapshot.totalClaimed,
            withdrawRequestedAt: snapshot.withdrawRequestedAt,
            refundNonce: snapshot.refundNonce,
            onchainSyncedAt: now,
          }),
      chargedCumulativeAmount: newCharged.toString(),
      signedMaxClaimable: voucher.maxClaimableAmount,
      signature: voucher.signature,
      lastRequestTimestamp: now,
    };
    outcome = { status: "committed", previous: base, current: updatedChannel };
    return updatedChannel;
  };

  let updateResult = await storage.updateChannel(channelId, current => charge(current, current));
  if (
    outcome?.status === "missing" &&
    snapshot &&
    (await inspectAdmission(scheme, channelId, requestContext?.pendingId)) === "self"
  ) {
    updateResult = await storage.updateChannel(channelId, current =>
      charge(current, current ?? snapshot),
    );
  }

  await scheme.clearPendingRequest(paymentPayload);

  if (outcome?.status === "missing") {
    return {
      abort: true,
      reason: Errors.ErrMissingChannel,
      message: "No channel record",
    };
  }

  if (outcome?.status === "cap_exceeded") {
    return {
      abort: true,
      reason: Errors.ErrChargeExceedsSignedCumulative,
      message: `Charged ${outcome.charged} exceeds signed max ${signedCap.toString()}`,
    };
  }

  if (updateResult.status !== "updated" || outcome?.status !== "committed") {
    return {
      abort: true,
      reason: Errors.ErrChannelBusy,
      message: "Concurrent request modified channel state",
    };
  }

  const skipExtra: BatchSettlementPaymentResponseExtra = {
    channelState: channelStateExtra(outcome.current, outcome.current.chargedCumulativeAmount),
    chargedAmount: requirements.amount,
  };

  return {
    skip: true,
    result: {
      success: true,
      payer: outcome.previous.channelConfig.payer.toLowerCase() as `0x${string}`,
      transaction: "",
      network: requirements.network,
      amount: "",
      extra: skipExtra,
    },
  };
}

/**
 * Enriches cooperative refund vouchers with facilitator settlement fields.
 *
 * @param scheme - Owning `BatchSettlementEvmScheme` instance for storage and signer access.
 * @param ctx - Settlement context for the current payment.
 * @returns Additive refund settlement fields, or nothing for non-refund payloads.
 */
export async function handleEnrichSettlementPayload(
  scheme: BatchSettlementEvmScheme,
  ctx: SettleContext,
): Promise<Record<string, unknown> | void> {
  const { paymentPayload, requirements } = ctx;
  const raw = paymentPayload.payload;
  if (!isBatchSettlementRefundPayload(raw)) {
    return;
  }

  const channelId = computeChannelId(raw.channelConfig, requirements.network);
  if (raw.voucher.channelId !== channelId) {
    throw new Error("refund channelId does not match channelConfig");
  }

  const requestContext = scheme.readRequestContext(paymentPayload);
  const snapshot = requestContext?.channelSnapshot;
  const stored = await scheme.getStorage().get(channelId);
  const pendingId = requestContext?.pendingId;
  const hold = await inspectAdmission(scheme, channelId, pendingId);
  if (!stored && snapshot && hold !== "self") {
    throw new Error(Errors.ErrMissingChannel);
  }
  const channel: Channel | undefined = snapshot
    ? {
        ...(stored ?? snapshot),
        ...snapshot,
        chargedCumulativeAmount:
          stored?.chargedCumulativeAmount ?? snapshot.chargedCumulativeAmount,
      }
    : stored;
  if (!channel) {
    throw new Error(Errors.ErrMissingChannel);
  }
  if (hold === "other") {
    throw new Error(Errors.ErrChannelBusy);
  }
  if (BigInt(raw.voucher.maxClaimableAmount) !== BigInt(channel.chargedCumulativeAmount)) {
    throw new Error(Errors.ErrCumulativeAmountMismatch);
  }
  if (raw.voucher.signature !== channel.signature) {
    throw new Error(Errors.ErrInvalidVoucherSignature);
  }

  const config = raw.channelConfig;

  const claimEntry: BatchSettlementVoucherClaim = {
    voucher: {
      channel: config,
      maxClaimableAmount: raw.voucher.maxClaimableAmount,
    },
    signature: raw.voucher.signature,
    totalClaimed: channel.chargedCumulativeAmount,
  };

  const remainder = BigInt(channel.balance) - BigInt(channel.chargedCumulativeAmount);
  if (remainder <= 0n) {
    throw new Error(Errors.ErrRefundNoBalance);
  }

  let refundAmountBig = remainder;
  if (raw.amount !== undefined) {
    if (!/^\d+$/.test(raw.amount)) {
      throw new Error(Errors.ErrRefundAmountInvalid);
    }
    const requested = BigInt(raw.amount);
    if (requested <= 0n) {
      throw new Error(Errors.ErrRefundAmountInvalid);
    }
    refundAmountBig = requested;
  }

  const refundAmount = refundAmountBig.toString();
  const nonce = String(channel.refundNonce ?? 0);

  const receiverAuthorizerSigner = scheme.getReceiverAuthorizerSigner();

  const refundAuthorizerSignature = receiverAuthorizerSigner
    ? await signRefund(
        receiverAuthorizerSigner,
        channelId as `0x${string}`,
        refundAmount,
        nonce,
        requirements.network,
      )
    : undefined;

  const claimAuthorizerSignature = receiverAuthorizerSigner
    ? await signClaimBatch(receiverAuthorizerSigner, [claimEntry], requirements.network)
    : undefined;

  scheme.rememberChannelSnapshot(paymentPayload, channel);

  return {
    ...(raw.amount === undefined ? { amount: refundAmount } : {}),
    refundNonce: nonce,
    claims: [claimEntry],
    refundAuthorizerSignature,
    claimAuthorizerSignature,
  };
}

/**
 * Lifecycle hook: runs after the facilitator settles a payment.
 *
 * Updates channel state to reflect the settlement outcome — adjusting charged amounts,
 * balances, and handling cooperative-refund cleanup (channel record deletion).
 *
 * @param scheme - Owning `BatchSettlementEvmScheme` instance for storage access.
 * @param ctx - Post-settle lifecycle context.
 * @param ctx.paymentPayload - Payment payload that was settled (possibly rewritten).
 * @param ctx.requirements - Requirements used for settlement.
 * @param ctx.result - Facilitator settle response.
 * @returns Resolves when session updates are complete (no return value).
 */
export async function handleAfterSettle(
  scheme: BatchSettlementEvmScheme,
  ctx: SettleResultContext,
): Promise<void> {
  const { paymentPayload, requirements, result } = ctx;
  if (!result.success) {
    return;
  }

  const raw = paymentPayload.payload;
  const storage = scheme.getStorage();

  if (isBatchSettlementRefundPayload(raw)) {
    const channelId = computeChannelId(raw.channelConfig, requirements.network);
    const pendingId = scheme.readRequestContext(paymentPayload)?.pendingId;
    const now = Date.now();

    const snapshot = parseRefundSettlementSnapshot(result.extra);
    const recovered = scheme.readRequestContext(paymentPayload)?.channelSnapshot;
    const hold = await inspectAdmission(scheme, channelId, pendingId);
    if (hold === "other") {
      throw new Error(Errors.ErrChannelBusy);
    }
    const updateResult = await storage.updateChannel(channelId, current => {
      const existing = current ?? (hold === "self" ? recovered : undefined);
      if (!existing) {
        return current;
      }
      if (BigInt(snapshot.balance) <= BigInt(existing.chargedCumulativeAmount)) {
        return undefined;
      }
      return {
        ...existing,
        ...snapshot,
        onchainSyncedAt: now,
        lastRequestTimestamp: now,
      };
    });
    if (updateResult.status === "unchanged") {
      throw new Error(Errors.ErrChannelBusy);
    }
    await scheme.releasePendingRequest(paymentPayload);
    if (!updateResult.channel) {
      return;
    }
    return;
  }

  if (isBatchSettlementVoucherPayload(raw)) {
    return;
  }

  if (isBatchSettlementDepositPayload(raw)) {
    const channelId = raw.voucher.channelId;
    const pendingId = scheme.readRequestContext(paymentPayload)?.pendingId;
    const ex = result.extra ?? {};
    const channelState = readChannelStateExtra(ex);
    const config = raw.channelConfig;
    const signedMaxClaimable = raw.voucher.maxClaimableAmount;
    const now = Date.now();

    const hold = await inspectAdmission(scheme, channelId, pendingId);
    if (hold === "other") {
      throw new Error(Errors.ErrChannelBusy);
    }
    const recovered = scheme.readRequestContext(paymentPayload)?.channelSnapshot;
    const updateResult = await storage.updateChannel(channelId, current => {
      const existing = current ?? (hold === "self" ? recovered : undefined);
      if (!existing) {
        return current;
      }
      const chargedActual = (
        BigInt(existing.chargedCumulativeAmount) + BigInt(requirements.amount)
      ).toString();
      return {
        channelId,
        channelConfig: config,
        chargedCumulativeAmount: chargedActual,
        signedMaxClaimable,
        signature: raw.voucher.signature,
        balance: readExtraString(channelState, "balance", existing.balance),
        totalClaimed: readExtraString(channelState, "totalClaimed", existing.totalClaimed),
        withdrawRequestedAt: readExtraNumber(
          channelState,
          "withdrawRequestedAt",
          existing.withdrawRequestedAt,
        ),
        refundNonce: readExtraNumber(channelState, "refundNonce", existing.refundNonce),
        onchainSyncedAt: now,
        lastRequestTimestamp: now,
      };
    });
    if (updateResult.status === "updated" && updateResult.channel) {
      scheme.rememberChannelSnapshot(paymentPayload, updateResult.channel);
      await scheme.releasePendingRequest(paymentPayload);
      return;
    }
    throw new Error(Errors.ErrChannelBusy);
  }
}

/**
 * Cleanup hook: clears this request's reservation after settlement throws.
 *
 * @param scheme - Owning `BatchSettlementEvmScheme` instance.
 * @param ctx - Settle failure context for the current payment.
 */
export async function handleSettleFailure(
  scheme: BatchSettlementEvmScheme,
  ctx: SettleFailureContext,
): Promise<void> {
  await scheme.clearPendingRequest(ctx.paymentPayload);
}

/**
 * Supplies server-owned settlement response fields from the channel snapshot.
 *
 * @param scheme - Owning `BatchSettlementEvmScheme` instance for snapshot access.
 * @param ctx - Settlement result context for the current payment.
 * @returns Additive response extra fields, or nothing when no snapshot exists.
 */
export async function handleEnrichSettlementResponse(
  scheme: BatchSettlementEvmScheme,
  ctx: SettleResultContext,
): Promise<Record<string, unknown> | void> {
  const raw = ctx.paymentPayload.payload;
  if (isBatchSettlementVoucherPayload(raw)) {
    return;
  }

  const channel = scheme.takeChannelSnapshot(ctx.paymentPayload);
  if (!channel) {
    return;
  }

  if (isBatchSettlementRefundPayload(raw)) {
    return {
      channelState: {
        chargedCumulativeAmount: channel.chargedCumulativeAmount,
      },
    };
  }

  if (isBatchSettlementDepositPayload(raw)) {
    return {
      channelState: {
        chargedCumulativeAmount: channel.chargedCumulativeAmount,
      },
      chargedAmount: ctx.requirements.amount,
    };
  }
  return {
    channelState: {
      chargedCumulativeAmount: channel.chargedCumulativeAmount,
    },
  };
}
